app/lib/package/backend.dart

```diff
@@ -832,28 +849,47 @@ class PackageBackend {
       }
 
       final pubspec = Pubspec.fromYaml(archive.pubspecContent!);
-      await _verifyPackageName(
-        name: pubspec.name,
-        user: user,
-      );
+      await _verifyPackageName(name: pubspec.name, user: user);
 
-      //
+      // Check if new packages are allowed to be uploaded.
       if (restriction == UploadRestrictionStatus.onlyUpdates &&
           !(await isPackageVisible(pubspec.name))) {
         throw PackageRejectedException.uploadRestricted();
       }
 
+      // Check version format.
       final versionString = canonicalizeVersion(pubspec.nonCanonicalVersion);
       if (versionString == null) {
         throw InvalidInputException.canonicalizeVersionError(
             pubspec.nonCanonicalVersion);
       }
 
+      // Check canonical archive.
+      final canonicalArchivePath =
+          tarballObjectName(pubspec.name, versionString);
+      final canonicalArchiveInfo =
+          await _canonicalBucket.tryInfo(canonicalArchivePath);
+      if (canonicalArchiveInfo != null) {
+        final file = File(filename);
+        if (canonicalArchiveInfo.length != await file.length()) {
+          throw PackageRejectedException.versionExists(
+              pubspec.name, versionString);
+        }
```

Comment:
Isn't this covered by `byteToByteEquals` ?

I think it's extremely rare that this happens, so we don't have to optimize for cases where the length does match.

Reply:
This is a cheap test that gets a fast-path rejection without actually fetching the bytes from the bucket. (E.g. if a client retries an upload too much, we will spend fewer resources on the validation and rejection...)